[WIP] Client batching

batiati opened the pull request:
## This PR
This is a take on #489.

**UPDATED**

Overview:

```zig
var batch = try client.get_batch(
    operation,
    @intCast(u32, payload.len),
);

// Write/serialize the payload to batch.slice():
std.mem.copy(u8, batch.slice(), payload);

client.submit_batch(
    user_data,
    on_complete,
    batch,
);
```

#### 1. Errors returned on `get_batch`.

Previously `TooManyOutstandingRequests` was notified via callback during `request`, but now we may have to check during `get_batch` if there is room for more requests.
Since `BatchLogical` is short-lived and the caller isn't responsible for acquiring/releasing a `message` anymore, they can't handle this error correctly via callback.

Also, `get_batch` can fail if the message size is invalid or too large.

#### 2. Functions `get_batch` + `submit_batch` replaces `request`.
We must guarantee that no operation can ever fail between `get_batch` and `submit_batch`.
Once `get_batch` returns, there is no API to cancel a batch, and it will leak if it wasn't submitted.

Should we add a function for that?
Example:

```zig
var batch = try client.get_batch(
    operation,
    @intCast(u32, payload.len),
);

// Cancels this batch in case of error:
errdefer batch.cancel();

// Write can fail here:
try decode(payload, batch.slice());

client.submit_batch(
    user_data,
    on_complete,
    batch,
);
```

#### 3. Tests and benchmarks that take advantage of batching.

All existing tests are passing, and new tests for demuxing batches were introduced. However, we don't have yet any specific test or benchmark taking advantage of batching.

#### 4. VOPR is currently broken

The `Simulator` ignores the regular callback, using a _hack_ function pointer `on_reply_callback` in order to inspect the messages with more details. Unfortunately, we can't reply with the demuxed data because this function expects the physical batch.

The simplest way to solve that is not allowing the Simulator to use logical batches. Having 1:1 batches/messages keeps `on_reply_callback` with the same semantics.
Or we can make the Simulator aware of logical batches, so it can correctly handle scenarios where a single message actually contains the reply of multiple requests.

## Pre-merge checklist

Performance:

* [X] Compare `zig benchmark` on linux before and after this pr.

Demuxing has no effect on 1:1 batch/message:

``` sh
# benchmark results before
Benchmarking...
1224 batches in 124.27 s
load offered = 1000000 tx/s
load accepted = 80471 tx/s
batch latency p00 = 0 ms
batch latency p10 = 34 ms
batch latency p20 = 37 ms
batch latency p30 = 39 ms
batch latency p40 = 41 ms
batch latency p50 = 47 ms
batch latency p60 = 50 ms
batch latency p70 = 53 ms
batch latency p80 = 58 ms
batch latency p90 = 67 ms
batch latency p100 = 4642 ms
transfer latency p00 = 0 ms
transfer latency p10 = 4467 ms
transfer latency p20 = 9309 ms
transfer latency p30 = 21764 ms
transfer latency p40 = 35420 ms
transfer latency p50 = 47266 ms
transfer latency p60 = 59682 ms
transfer latency p70 = 73517 ms
transfer latency p80 = 86291 ms
transfer latency p90 = 100383 ms
transfer latency p100 = 114277 ms

# benchmark results after
Benchmarking...
1224 batches in 122.32 s
load offered = 1000000 tx/s
load accepted = 81753 tx/s
batch latency p00 = 0 ms
batch latency p10 = 34 ms
batch latency p20 = 37 ms
batch latency p30 = 39 ms
batch latency p40 = 41 ms
batch latency p50 = 47 ms
batch latency p60 = 50 ms
batch latency p70 = 52 ms
batch latency p80 = 55 ms
batch latency p90 = 65 ms
batch latency p100 = 4059 ms
transfer latency p00 = 0 ms
transfer latency p10 = 4452 ms
transfer latency p20 = 11737 ms
transfer latency p30 = 24702 ms
transfer latency p40 = 34710 ms
transfer latency p50 = 46687 ms
transfer latency p60 = 59326 ms
transfer latency p70 = 72488 ms
transfer latency p80 = 85299 ms
transfer latency p90 = 98405 ms
transfer latency p100 = 112329 ms
```

batiati committed:
WIP Client batching: Initial/experimental API

batiati committed:
Comments typo

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -19,21 +19,130 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
         pub const StateMachine = StateMachine_;
 
         pub const Error = error{
+            BatchSizeZero,
```
I think it should be possible more generally for an operation to allow an empty body. (Even though none of the operations besides `register` currently do so). Maybe `StateMachine` should have a set of `batch_min` constants (mirroring `batch_max`).

batiati replied:
In a hypothetical case of empty body, would it be correct to batch multiple logical batches into a single request? Since there is no message body, I think this particular case would require multiple physical requests as well to keep the intended semantics. Otherwise, the server will see only one message, not multiple.

Let's move all message body validations and the decision whether an operation can be batched or not to the `StateMachine`. 👍

jorangreef replied:
> Since there is no message body, I think this particular case would require multiple physical requests as well to keep the intended semantics.

Agreed. I think that an empty body should not be batched.

> Let's move all message body validations and the decision whether an operation can be batched or not to the StateMachine.

We do want to keep as much as possible out of the state machine, to make it easier for users to create custom state machines. If there's anything that we can possibly validate outside of the state machine, then we should do that, to keep complexity contained within VSR.

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -19,21 +19,130 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
+        pub const BatchLogical = struct {
+            operation: StateMachine.Operation,
+            size: u32,
+            batch_physical: union(enum) {
```
Why not just always use `{ *Request, offset }`, and eagerly add the empty `Request` to the queue when `get_batch()` is called?

batiati replied:
I found it less error-prone not having to check for an eager/incomplete request, but rather keeping the check if the request queue was empty (eg `if (was_empty) self.send_request_for_the_first_time(message);`)

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -373,19 +579,90 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
+        /// (3,x),(5,x)  | (2,10) | (1,x),(3,x)
+        /// ()           | (12,8) | ()
+        /// (20,x)       | (20,1) | (0,x)
+        fn decode_reply(self: *Self, comptime operation: StateMachine.Operation, demux_list: DemuxList, reply: *Message) void {
```
This logic will (in part) need to move into `StateMachine`, since the `result.index` mapping is statemachine-specific logic.

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -19,21 +19,130 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
+
+        pub const Request = struct {
+            // Null iif operation=register.
+            demux_list: ?DemuxList,
```
Rather than a pool of Demux, and a linked list on each request, why not just put a BoundedArray on each Request? I think that would save a little memory since we wouldn't need the `next` pointers anymore.

batiati replied:
Do you mean by not having an adjustable upper bound? If we choose to allow up to 8191 logical batches for each physical request, it will save us ~2MB of memory.
On the other hand, a pool + linked list can be helpful if we choose to share a limited number of logical batches across all requests (let's say, something like 10k logical batches would require only 400KB of memory).

sentientwaffle replied:
:+1: Good point, an adjustable/configurable upper-bound is better, since clients can tune it based on their own average logical batch size.

jorangreef commented on src/vsr/client.zig:
```diff
@@ -19,21 +19,130 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
         pub const Error = error{
+            BatchSizeZero,
+            TooMuchData,
             TooManyOutstandingRequests,
```
Let's use our most-significant bit ordering here for these names, e.g. `RequestLimitExceeeded` or `RequestBodyExceeded` or variants on that style.

jorangreef replied:
Not that my suggestions are what we should go with, but just to give an example of the MSB order.

batiati committed:
[WIP] Client batching

- Move batching/demux logic to StateMachine.
- Better names and code organization.
- Tb_client changes

batiati commented:
In the last commit, I implemented enough to run the Java client tests (including integration tests).
So, I could run a simple benchmark to measure the creation of 1_000 async transfers (one per task)

**Before:**
```
Time for 1000 transactions: 3022 ms
```

**After (with batching)**
```
Time for 1000 transactions: 68 ms
```

Source code:

```java
final int tasks_qty = 1_000;

try (var client = new Client(0, new String[] {Server.TB_PORT}, tasks_qty)) {
    var errors = client.createAccounts(accounts);
    assertTrue(errors.getLength() == 0);

    var now = System.currentTimeMillis();
    final var tasks = new CompletableFuture[tasks_qty];
    for (int i = 0; i < tasks_qty; i++) {

        final var transfers = new TransferBatch(1);
        transfers.add();

        transfers.setId(UInt128.asBytes(UUID.randomUUID()));
        transfers.setCreditAccountId(account1Id);
        transfers.setDebitAccountId(account2Id);
        transfers.setLedger(720);
        transfers.setCode((short) 1);
        transfers.setAmount(100);

        // Starting async batch
        tasks[i] = client.createTransfersAsync(transfers);
    }

    // Wait for all threads
    for (int i = 0; i < tasks_qty; i++) {
        @SuppressWarnings("unchecked")
        final var future = (CompletableFuture<CreateTransferResultBatch>) tasks[i];
        final var result = future.get();
        assertEquals(0, result.getLength());
    }

    var elapsed = System.currentTimeMillis() - now;
    System.out.printf("Time for %d transactions: %d ms\n", tasks_qty, elapsed);
}
```

batiati committed:
Merge from main

batiati committed:
Client batching: Demux tests

batiati committed:
Client batching: Refactoring call sites

batiati committed:
Merge branch 'main' into batiati-client-batch

batiati committed:
Client batching: more API rework

- Constant names refactor
- Hide types not exposed by the public API
- Update call-sites

batiati committed:
Merge branch 'main' into batiati-client-batch

batiati committed:
Merge changes

batiati committed:
[Client batching] More API rework

batiati committed:
[Client batching] More call site changes

batiati committed:
Merge branch 'main' into batiati-client-batch

batiati committed:
[Client batching] Fix fn signature

batiati committed:
[Client batching] Node client

batiati committed:
Merge branch 'main' into batiati-client-batch

batiati committed:
zig fmt

batiati committed:
[Client batching] Echo client

batiati commented:
The new batching API was updated in all call sites, and all tests are passing now.
For more details, please see the EDITs on this PR text.

The next step is to implement specific tests and benchmarks for batching.

batiati committed:
Merge branch 'main' into batiati-client-batch
